### PR TITLE
Upgrade to pnpm

### PR DESCRIPTION
**Pull request by technophile-04**
## Description

Updated the pkgManager to pnpm for trial as discussed in #419 

~For deploying contract you need to use `pnpm hardhat:deploy` instead of just `pnpm deploy` because [`deploy`](https://pnpm.io/cli/deploy) seem reserved keyword~
~Will try to find if there exist any solution to above problem 🙌....But I think its ready to test and give a trial~ 

This is solved please check https://github.com/scaffold-eth/scaffold-eth-2/pull/444#issuecomment-1667241333 🙌

### CODE REVIEW

**Comment by rin-st** [comment]
I tried base scripts `chain`, `hardhat:deploy`, `start`, `vercel` and some others and everything works as expected 👍 . `pnpm hardhat:deploy` is a little bit longer than it was, but it becomes more clear what is deployed, at least for newbies.

File: package.json
Comment (rin-st) [inline package.json:31]: why 8.3.1, not 8.6.10?

**Comment by rin-st** [comment]
and it shows me error when I use `ctrl + c` in `pnpm chain` terminal. 
image

As I understand it doesn't affect anything. Fast googling didn't help me but we need to try to get rid of it.

File: package.json
Comment (carletex) [inline package.json:15]: Why do we need the prefix `hardhat` here?

I think we want to keep chain / deploy / start as it was.

Comment (carletex) [reply package.json:15]: Oh, I read the description now... sorry. haha

That is going to be a though one....

File: pnpm-workspace.yaml
Comment (carletex) [inline pnpm-workspace.yaml:3]: Could this also be `'packages/*'`? Thinking about the cli :D

File: README.md
Comment (carletex) [inline README.md:31]: Was removing `Git` an accident or you don't want it there? haha

File: package.json
Comment (carletex) [reply package.json:31]: Got this when running `pnpm install`

image

**Comment by carletex** [comment]
Great job @technophile-04! Thank you for putting this together.

Tested it and everything works great.

> For deploying contract you need to use pnpm hardhat:deploy instead of just pnpm deploy because [deploy](https://pnpm.io/cli/deploy) seem reserved keyword

Damn, that's going to be a hard pill to swallow for some people haha. The `deploy` command is the most used one + we are switching to pnpm: `yarn deploy` => `pnpm hardhat:deploy` is going to be hard to sell haha. As Rinat said, probably for newcomers it doesn't matter. We could also do `contract:deploy`.

**Comment by Pabl0cks** [comment]
It's working great on Windows, GJ @technophile-04 !!

**Comment by technophile-04** [comment]
> As I understand it doesn't affect anything. Fast googling didn't help me but we need to try to get rid of it.

> pnpm is less friendly than yarn when you write a non existing command: 

I have been playing around with some configs mentioned here -> https://pnpm.io/npmrc#cli-settings but nothing works :( 

---  

I have also noticed a bug i.e patch is not applied, because if you deploy two contracts you get : 

| YourContarct is selected but not highlight | hydration warning in console | 
| ------------------------------------------| -----------------------------| 
|Screenshot 2023-07-28 at 6 58 53 PM | Screenshot 2023-07-28 at 6 58 31 PM | 

This signifies that patch is not working checkout https://github.com/scaffold-eth/scaffold-eth-2/issues/453#issuecomment-1655732858 for more description 

Having a look at it 🙌

File: package.json
Comment (technophile-04) [reply package.json:11]: > "preinstall": "npx only-allow pnpm",

Added at https://github.com/scaffold-eth/scaffold-eth-2/pull/444/commits/e720b86d015bb3316de070a7ac9c6eb68ab8d45a 🙌

File: README.md
Comment (technophile-04) [reply README.md:31]: fixed at https://github.com/scaffold-eth/scaffold-eth-2/pull/444/commits/ec68f7075de0e07e2757e5556e24ff426ee079e7

File: pnpm-workspace.yaml
Comment (technophile-04) [reply pnpm-workspace.yaml]: Yup fixed at https://github.com/scaffold-eth/scaffold-eth-2/pull/444/commits/ba393c0061a698343590b66ae34994ed13a4ae81 🙌

File: package.json
Comment (technophile-04) [reply package.json]: Updated at [d148eb0](https://github.com/scaffold-eth/scaffold-eth-2/pull/444/commits/d148eb050c6ac96892db959b700fda9b6a8d452e)

**Comment by technophile-04** [comment]
Update I got it working in [pnpm-pathc](https://github.com/scaffold-eth/scaffold-eth-2/tree/fix/pnpm-pathc) branch, but to get it working I need to do some manual work :(, and also there if you run any pnpm command you always get this warning : 
> The field "pnpm" was found in `/Users/..../scaffold-eth-2/packages/next
js/package.json.` This will not take effect. You should configure "pnpm" at the root of the workspace instead. 

I have created an issue on pnpm -> https://github.com/pnpm/pnpm/issues/6885 

Let's see what they have to say if there is some direct way to do this otherwise we merge [pnpm-pathc](https://github.com/scaffold-eth/scaffold-eth-2/tree/fix/pnpm-pathc) into this branch

I will keep tinkering around it 🙌

File: package.json
Comment (technophile-04) [reply package.json]: Thanks to this review comment  !!, pnpm now shows friendly messages 😁 

Screenshot 2023-08-07 at 10 09 36 AM


| Yarn | Pnpm | 
| ----- | ----- | 
| Screenshot 2023-08-07 at 10 14 40 AM |  Screenshot 2023-08-07 at 10 14 12 AM | 


Pnpm also suggests the closest command unlike `yarn` :) 


Thanks to  this PR https://github.com/pnpm/pnpm/pull/6887  too 🙌

**Comment by technophile-04** [comment]
## Some updates : 

### Regarding `deploy` command : 

To run scripts from package.json you need to use `run` similar to `npm` : 

```
pnpm run script_name
```

Eg : 

```
pnpm run deploy 
```

Yeah :), we can use our old `deploy` command to deploy contracts was fixed at [714a6e8](https://github.com/scaffold-eth/scaffold-eth-2/pull/444/commits/714a6e85f8b78d60ea4d7688d291311065d48877)

And to run binaries inside `node_modules` eg : `npx hardhat node` instead of `npx` we directly use `pnpm hardhat node`

Rinna'ts comment : 
> and it shows me an error when I use `ctrl + c` in `pnpm chain` terminal. 
> image
> As I understand it doesn't affect anything. Fast googling didn't help me but we need to try to get rid of it.

If you do `cd packages/hardhat`, and then run `pnpm hardhat node --network hardhat --no-deploy `  now if you hit `ctrl +c` it will close nicely 

This is not a big deal as it doesn't affect anything but will try to find if there is some workaround 🙌

---

### Regarding patch being not applied as discussed in https://github.com/scaffold-eth/scaffold-eth-2/pull/444#issuecomment-1657750641 : 

It seems there was some problem from `pnpm` side, and it was solved in recent version 🙌

I have also updated this at  [073ce50](https://github.com/scaffold-eth/scaffold-eth-2/pull/444/commits/073ce50afc8ff6c606539b8a612c2a1480ff77d0) with new patch fix and it works nicely now 

--- 


### Blocking thing : 

Checkout this issue I have created : https://github.com/pnpm/pnpm/issues/6890

I think this is the only issue which is blocking us now, once this is solve we can go through test once again and it will ready to merge 🙌

**Comment by technophile-04** [comment]
Hey guys this is ready for testing and review 🙌 

Just one thing I noticed on subsequent installs on macOS 

Pnpm takes `2.3s` and Yarn takes `665ms` 

> pnpm install might be slower on macOS because pnpm uses hard links on macOS and Yarn copies files. Copying is faster but uses more disk space. Copy-on-write copies would be the best (which we use on Linux filesystems that support it) but on macOS we can't use it because Node.js doesn't allow it (https://github.com/pnpm/pnpm/issues/5001) 
~ It was mentioned here: https://github.com/pnpm/pnpm/issues/6447

But would love it if others could test it as well on Windows & Linux

**Comment by damianmarti** [comment]
@technophile-04 I successfully run **pnpm install** (a new version is online 8.7.1) and I was able to run all the command but **pnpm hardhat:deploy**

I got:

 `
ERR_PNPM_RECURSIVE_EXEC_FIRST_FAIL Command failed with ENOENT: hardhat:deploy

spawn hardhat:deploy ENOENT
` 

**pnpm hardhat:test** works fine and all the tests are passing.

I'm on Ubuntu 20.04.

**Comment by damianmarti** [comment]
> Hey guys this is ready for testing and review 🙌
> 
> Just one thing I noticed on subsequent installs on macOS
> 
> Pnpm takes `2.3s` and Yarn takes `665ms`
> 
> > pnpm install might be slower on macOS because pnpm uses hard links on macOS and Yarn copies files. Copying is faster but uses more disk space. Copy-on-write copies would be the best (which we use on Linux filesystems that support it) but on macOS we can't use it because Node.js doesn't allow it ([pnpm/pnpm#5001](https://github.com/pnpm/pnpm/issues/5001))
> > ~ It was mentioned here: [pnpm/pnpm#6447](https://github.com/pnpm/pnpm/issues/6447)
> 
> But would love it if others could test it as well on Windows & Linux

How are you measuring these times "Pnpm takes 2.3s and Yarn takes 665ms"?

**Comment by technophile-04** [comment]
> @technophile-04 I successfully run pnpm install (a new version is online 8.7.1) and I was able to run all the command but pnpm hardhat:deploy ...

Actually to deploy, we should use `pnpm run deploy` actually I forgot to update the first PR comment my bad will update it but here is the detailed explanation  https://github.com/scaffold-eth/scaffold-eth-2/pull/444#issuecomment-1667241333

Could you please try `pnpm run deploy` once and also make sure you append `run` while scripts from package.json ? 

Also one thing regarding the error message you got, could you please do `pnpm -v` and see if your system `pnpm` version is `8.7.1` ? Because ideally the error message should have been a bit friendly : 
Screenshot 2023-09-07 at 1 30 23 PM

--- 

> How are you measuring these times "Pnpm takes 2.3s and Yarn takes 665ms"?

Actually my terminal outputs "Done with ..."

| Pnpm | Yarn | 
| ------| ------|
| Screenshot 2023-09-07 at 1 34 09 PM |  Screenshot 2023-09-07 at 1 34 55 PM | 

Maybe other's could append `time` before running command eg: `time pnpm install` to get actual time taken by process 

Screenshot 2023-09-07 at 1 38 24 PM

Also, this is on subsequent install with all the cache present

**Comment by Pabl0cks** [comment]
> But would love it if others could test it as well on Windows & Linux

I'm playing with pnpm on Windows, everything seems to be working fine for now, except **pnpm hardhat:test**. It's throwing this exception:  ERR_PNPM_RECURSIVE_RUN_FIRST_FAIL  @se-2/hardhat@0.0.1 test: `REPORT_GAS=true hardhat test --network hardhat`

If I delete `REPORT_GAS=true` from test script, it works fine. 

---

On my windows desktop it takes 10s to pnpm install on subsequent installs, vs 2.5s from yarn install. So it seems the same 4:1 ratio. 

---

Will play a bit more with my SE-2 build and report back if I find anything. GJ @technophile-04 🙌🙌

**Comment by damianmarti** [comment]
> pnpm hardhat:deploy

It is working now! Thanks!!

I had 8.5.1 and now I updated to 8.7.4.

It seems like pnpm is a little slower than yarn but I wouldn't bother about it...

**Review by damianmarti — APPROVED** [review]
Great work @technophile-04 !!

We should check if we need to update something on the documentation website...

**Comment by carletex** [comment]
Amazing job @technophile-04 it works great!! And thank you all for testing.

So:

**PROS**
- allows us to have individual .locks files per package.
- Fancy new tool that people like more, now? :D Not a fan of this PRO tho haha

**CONS**
- Remove OG `yarn <script>`. Not it's `pnpm run <action>`.
- Slower. (for me too) Like Damu said, not a super big deal.

What do you all think?

**Comment by damianmarti** [comment]
> Amazing job @technophile-04 it works great!! And thank you all for testing.
> 
> So:
> 
> **PROS**
> 
> * allows us to have individual .locks files per package.
> * Fancy new tool that people like more, now? :D Not a fan of this PRO tho haha
> 
> **CONS**
> 
> * Remove OG `yarn <script>`. Not it's `pnpm run <action>`.
> * Slower. (for me too) Like Damu said, not a super big deal.
> 
> What do you all think?

I think the really important issue is the advantage of having individual .locks files (the other PROS and CONS are not too relevant, I think). 

So I vote to go with pnpm! ;-)

**Comment by carletex** [comment]
> the other PROS and CONS are not too relevant, I think

This is very subjective, that's why I think it's important for us to vote :D

Austin is neutral:
image

I think I'm 51% pnpm / 49 % yarn :D

If we want to go with PNMP, we'll need to sync:
1. Merge this PR
2. Update the SE-docs site
3. Merge to CLI + publish. (We should also test it well)

We should probably have everything prepared and push the red button :red_circle: when ready.

**Comment by carletex** [comment]
In case we move to pnpm: PR in the SE-2 docs repo: https://github.com/scaffold-eth/se2-docs/pull/20

It's a bit weird to use `pnpm <script>` for all the scripts except for `deploy`. Should we recommend (in the docs) to use `pnpm run` for all the scripts, or do you think is it good as it is?

File: README.md
Comment (carletex) [inline README.md:23]: I think this should point to https://pnpm.io/installation

Also maybe "pnpm" lowercase? 

lol, they are trolling in their site, when you reload it displays in a different case:

image
image

**Comment by technophile-04** [comment]
>  fine for now, except pnpm hardhat:test. It's throwing this exception:

> On my windows desktop it takes 10s to pnpm install on subsequent installs, vs 2.5s from yarn install.

This is strange :(, @Pabl0cks could you please just check your `pnpm` version and making sure its above `>= 8.7.1` because I had this similar 4:1 ratio before version `8.7.1` because of https://github.com/pnpm/pnpm/issues/6890 but its fixed now 

Also regarding `pnpm hardhat:test` could you just try using `pnpm run hardhat:test` and also share an SS of full output 🙌


----

I agree with you guys ! 

I only vote for `pnpm` because it will make installation via CLI a lot faster and consistent, as we will already be having "pnpm-lock.yam" inside each extension / package right ? (LOL the main motive for me to create this PR was this reason) but if we are not able to include `pnpm-lock.yaml` inside each package then I don't think its worth it  (Please correct me If I missed something or lol I am wrong path or missed some other PRO's)

Regarding #419 this could be solved without using `pnpm` by uploading the whole Monorepo to vercel (which in turn will contain `yarn.lock`) but people need to mention "./packages/nextjs" when they run "yarn vercel" as mentioned in https://github.com/scaffold-eth/scaffold-eth-2/pull/231#issuecomment-1471632149 which might be ok then typing "pnpm run" every time  


> We should probably have everything prepared and push the red button 🔴 when ready.

Lol yup, I was wondering whether we should create another PR parallel / test locally at CLI branch utilizing pnpm to install dependency and see if everything works nicely there and then we can merge this PR once we are sure ?   

Tysm guys!!

**Comment by Pabl0cks** [comment]
> This is strange :(, @Pabl0cks could you please just check your pnpm version and making sure its above >= 8.7.1 because I had this similar 4:1 ratio before version 8.7.1 because of https://github.com/pnpm/pnpm/issues/6890 but its fixed now
>
> Also regarding pnpm hardhat:test could you just try using pnpm run hardhat:test and also share an SS of full output 🙌

I had 8.6.10. Updated to 8.7.4 and installation went down to 8.6 seconds 🙌. 

I still get the `REPORT_GAS=true` error, and I've noticed a error during the install, don't know if it can affect my `pnpm hardhat:test` results. Here is the SS with the 3 commands:

image

Here the SS using PowerShell instead:

image

File: README.md
Comment (carletex) [reply README.md]: Fixed the link here: https://github.com/scaffold-eth/scaffold-eth-2/pull/444/commits/a792ecaeb3794cac07d0d2e88446f0b9f01a375f

**Comment by carletex** [comment]
@Pabl0cks I think Windows doesn't recognize env vars like that on the `package.json`.

Can you check if you get the same error when running `yarn hardhat:test` on `main`? We might need to use something like https://www.npmjs.com/package/cross-env

**Comment by technophile-04** [comment]
> It's a bit weird to use `pnpm <script>` for all the scripts except for deploy. Should we recommend (in the docs) to use `pnpm run` for all the scripts, or do you think is it good as it is?

I would vote for using `pnpm run <script>` just to keep things consistent and also pnpm [suggests](https://pnpm.io/cli/run) it to run that way. 

Also noticed `pnpm run` gives a slightly correct `ERR_MESAAGE` output if you misspell : 
Screenshot 2023-09-14 at 8 04 23 PM

Lol but yeah I am kind of 51% and 49% so anything works 🙌

**Comment by Pabl0cks** [comment]
> Can you check if you get the same error when running `yarn hardhat:test` on `main`? 

Just tried a fresh `scaffold-eth-2` and `yarn hardhat:test` works fine on `main`. Here is the sshot:

image

**Comment by rin-st** [comment]
@technophile-04 since you [fixed](https://github.com/scaffold-eth/scaffold-eth-2/pull/499) `wagmi/viem` types, original problem disappeared.

The only problem with yarn for me is this 
image

But if we move that simple type to our project everything works again with `yarn` without locking wagmi/viem versions. Am I missing something?

**Comment by technophile-04** [comment]
> The only problem with yarn for me is this image
> 
> But if we move that simple type to our project everything works again with `yarn` without locking wagmi/viem versions.

Tysm Rinat, just created PR at #541 
 
> So it seems we don't need lockfiles anymore. Am I missing something?

Do you, mean to say separate lockfiles? 
 
I think although the `viem/wagmi` issue is solved but it might with other libraries that use installs when using SE-2 for example checkout => https://github.com/scaffold-eth/se-2-challenges/pull/91 (this can solved as mentioned in https://github.com/scaffold-eth/scaffold-eth-2/pull/444#issuecomment-1711647676 but it decreases DX while deploying to vercel )

also checkout Carlos comment here => https://github.com/scaffold-eth/scaffold-eth-2/issues/290#issuecomment-1714052355 🙌

**Comment by rin-st** [comment]
> Tysm Rinat, just created PR at https://github.com/scaffold-eth/scaffold-eth-2/pull/541

👍 ✅ 

> Do you, mean to say separate lockfiles?

Yep

> I think although the viem/wagmi issue is solved but it might with other libraries that use installs when using SE-2 for example checkout => https://github.com/scaffold-eth/se-2-challenges/pull/91

hm, agree, you're right. And I'm glad you're right )

> (this can solved as mentioned in https://github.com/scaffold-eth/scaffold-eth-2/pull/444#issuecomment-1711647676 but it decreases DX while deploying to vercel )

I tried it, doesn't work for me. It asks for additional packages like `node-gyp`. Didn't try to install it, I think it's not a good solution

❤️

**Review by rin-st — APPROVED** [review]

**Comment by technophile-04** [comment]
Can confirm that the error Pablo got https://github.com/scaffold-eth/scaffold-eth-2/pull/444#issuecomment-1712319461 here is with all Windows users, tried it on my friend's Windows machine and got the same error. 


After doing some digging I got it working on Windows by first running (needs to be run only once in life) : 

```shell
pnpm config set shell-emulator=true
```

after doing this `pnpm run hardhat:test` worked fine (@Pabl0cks could you please try this once ? )

**Reasoning :**

checkout description for [shell-emulator](https://pnpm.io/cli/run#shell-emulator) as pnpm says : 

> When true, pnpm will use a JavaScript implementation of a [bash-like shell](https://www.npmjs.com/package/@yarnpkg/shell) to execute scripts

Also, I think the reason why it works with `yarn` by default is because yarn handles it already internally check out this => [yarn/features/scripting/portable-shell](https://yarnpkg.com/features/scripting#portable-shell)

--- 

Just some thoughts while I was playing on Windows: 

Lol don't know why but I don't feel confident merging this PR (at least because of windows)
   - For some reason it took lot of time while doing install almost around 5m 6.6s (It might be due network condition but yarn install was comparatively a lot better)
   - On subsequest install was getting the same `cpu-feature` error as Pablo showed in his first image in https://github.com/scaffold-eth/scaffold-eth-2/pull/444#issuecomment-1712319461 (planning to debug this next 🙌)

**Comment by rin-st** [comment]
> After doing some digging I got it working on Windows by first running (needs to be run only once in life)

Since your [pnpm config is inside .npmrc file](https://pnpm.io/cli/config) you can just add it to npmrc [example](https://petermekhaeil.com/til/pnpm-shell-emulator/) and it will work. Could you check it?

With this option, it also works on mac. Installed all deps in 15.3 sec (Mac pro 2021)

**Comment by Pabl0cks** [comment]
> After doing some digging I got it working on Windows by first running (needs to be run only once in life) :
> 
> pnpm config set shell-emulator=true
> after doing this pnpm run hardhat:test worked fine (@Pabl0cks could you please try this once ? )

Yeah, it worked !! Great job @technophile-04 🙌

**Comment by carletex** [comment]
Some updates that might affect this PR here: https://github.com/scaffold-eth/scaffold-eth-2/pull/544#issuecomment-1737451534

**Comment by carletex** [comment]
As we discussed, closing this for now. Let's keep the branch tho!

**Pull request closed without merging**
